dut/drivers.wake

```diff
@@ -0,0 +1,188 @@
+  helper root Nil | JArray
+
+tuple DriverImplementation =
+  global IsCompatible: List String
+  global Sources:      List Path
+  global IncludeDirs:  List String
+
+global def makeDriverImplementation compatibleDevices driverSources includeDirs =
+  DriverImplementation compatibleDevices driverSources includeDirs
+
+target driverImpCompatibilityTrees Unit =
+  subscribe driverImplementations
+  | map (\p p → listToTree scmp p.getDriverImplementationIsCompatible)
+
+target getCompatibleDriverImplementation compatibleString =
+  def drivers = filter (compatibleString ∈ _.getPairSecond) Unit.driverImpCompatibilityTrees
+  match drivers
+    one, Nil = Some one.getPairFirst
+    Nil = None
+    _ =
+      def _ = printlnLevel logWarn "WARNING: More than one driver implementation available for device {compatibleString}. A random one will be chosen."
+      None
+
+global target getCompatibleLists dtsJSON =
+  def isDevice = match _
+    JObject fields = exists (matches `compatible` _.getPairFirst) fields
+    _ = False
+  def getLists =
+    _
+    | getJArray
+    | getOrElse Nil
+    | mapPartial (getJArray)
+  match dtsJSON.parseJSONFile
+    Pass dtsJValue =
+      dtsJValue
+      | (_ /../ isDevice)
+      | getJArray
+      | getOrElse Nil
+      | map (_ // `compatible`)
+      | mapPartial getJArray
+      | map getJString.mapPartial
+    Fail e =
+      def _ = printlnLevel logWarn "WARNING: Could not parse DTS JSON: {e.getErrorCause}"
+      Nil
+
+global def getDriverImplementationsForDUT dut =
+  def dts = dut.getDUTDTS
+  def omFile = dut.getDUTObjectModelFile
+  def compatibleLists = getCompatibleLists dut.getDUTDTSJSON
+  def getImpForCompatibleList compatibleList =
+    compatibleList
+    | mapPartial getCompatibleDriverImplementation
+    | head
+  mapPartial getImpForCompatibleList compatibleLists
+
+
+tuple BasicDriverPlan =
+  global DUHDocument:         Path
+  global OMFile:              Unit => Path
+  global VendorName:          String
+  global DeviceName:          String
+  global LocalMetalDirectory: String
+  global BSPDirectory:        String
+  global IsCompatible:        List String
+  global DUTName:             String
+  global OverwriteExisting:   Boolean
+  global BaseHeader:          Boolean
+  global BasicDrivers:        Boolean
+
+# create a BasicDriverPlan
+# duhFile => path to your duh file (device.json5)
+# OMFile => a function to produce the OMFile, takes Unit as an arg
+# VendorName => the vendor name used for the device
+# DeviceName => the name of the device
+# LocalMetalDirectory => the metal/drivers directory in the device's repository
+global def makeBasicDriverPlan duhFile omFile vendorName deviceName localMetalDirectory bspDirectory compat dutName =
+  BasicDriverPlan duhFile omFile vendorName deviceName localMetalDirectory bspDirectory compat dutName True True True
+
+# memoize this because trees in wake are slow
+target basicDriverCompatibilityTrees Unit =
```

Comment:
I've never properly learned how `target` works, since there's no documentation on it. Is it safe to declare a function as a `target` when it only accepts a `Unit` and within the body, there's a `subscribe`?

@albertchen-sifive, @jackkoenig, or @terpstra, can you explain how `target` works? Even better is if we could properly document it in `wake.git`.

Also, another I've been wondering is whether the order of items coming back from a `subscribe` is deterministic.

Reply:
I think it's fine to have a `subscribe` within a `target` since wake has to "elaborate" all the pub/subs at compile time in order to check for circular dependencies?

Reply:
target explanation https://github.com/sifive/wake/issues/124

I don't think the final implementation has changed from how it is described in the issue other than that target will panic if your "ignored" arguments differ for the same "key" arguments.

Reply:
Thanks, that's a very useful ticket to learn about `target`!